sdk/opa_test.go

```diff
@@ -0,0 +1,302 @@
+	}
+
+}
+
+func TestDecisionLogging(t *testing.T) {
+
+	ctx := context.Background()
+
+	server := sdktest.MustNewTestServer(
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": "package system\nmain = time.now_ns()",
+		}),
+	)
+
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/bundle.tar.gz"
+			}
+		},
+		"decision_logs": {
+			"console": true
+		}
+	}`, server.URL()))
+
+	testLogger := loggingtest.New()
+	opa, err := New(ctx, Config(config), Console(testLogger))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer opa.Stop(ctx)
+
+	// Verify that timestamp matches time.now_ns() value.
+	if _, err := opa.Decision(ctx, DecisionParams{
+		Now: time.Unix(0, 1619868194450288000).UTC(),
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	entries := testLogger.Entries()
+	exp := json.Number("1619868194450288000")
+
+	if len(entries) != 1 || entries[0].Fields["result"] != exp || entries[0].Fields["timestamp"] != "2021-05-01T11:23:14.450288Z" {
+		t.Fatalf("expected %v but got %v", exp, entries[0].Fields["result"])
+	}
+
+}
+
+func TestDiscovery(t *testing.T) {
+
+	ctx := context.Background()
+
+	server := sdktest.MustNewTestServer(
+		sdktest.MockBundle("/bundles/discovery.tar.gz", map[string]string{
+			"bundles.rego": `
+				package bundles
+
+				test := {"resource": "/bundles/bundle.tar.gz"}
+			`,
+		}),
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": `
+				package system
+
+				main = 7
+			`,
+		}),
+	)
+
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"discovery": {
+			"resource": "/bundles/discovery.tar.gz"
+		}
+	}`, server.URL()))
+
+	opa, err := New(ctx, Config(config))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer opa.Stop(ctx)
+
+	exp := json.Number("7")
+
+	if result, err := opa.Decision(ctx, DecisionParams{}); err != nil {
+		t.Fatal(err)
+	} else if result.Result != exp {
+		t.Fatalf("expected %v but got %v", exp, result.Result)
+	}
+
+}
+
+func TestAsync(t *testing.T) {
+
+	ctx := context.Background()
+
+	readyCh := make(chan struct{})
+
+	server := sdktest.MustNewTestServer(
+		sdktest.MockBundle("/bundles/bundle.tar.gz", map[string]string{
+			"main.rego": "package system\nmain = 7",
+		}),
+	)
+
+	defer server.Stop()
+
+	config := []byte(fmt.Sprintf(`{
+		"services": {
+			"test": {
+				"url": %q
+			}
+		},
+		"bundles": {
+			"test": {
+				"resource": "/bundles/bundle.tar.gz"
+			}
+		}
+	}`, server.URL()))
+
+	opa, err := New(ctx, Config(config), Async(readyCh))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if result, err := opa.Decision(ctx, DecisionParams{}); !IsUndefinedErr(err) {
+		t.Fatal("expected undefined error but got", result, "err:", err)
+	}
+
+	defer opa.Stop(ctx)
+
+	select {
+	case <-readyCh:
+		// ready
+	}
```

Comment:
```suggestion
<-readyCh:  // ready
```

unless I'm missing something 🤔

Reply:
Not missing anything. I was on autopilot.